homeassistant/components/group/entity.py

```diff
@@ -287,23 +288,49 @@ def _set_tracked(self, entity_ids: Collection[str] | None) -> None:
         if not entity_ids:
             self.tracking = ()
             self.trackable = ()
+            self.single_state_type_key = None
             return
 
         registry: GroupIntegrationRegistry = self.hass.data[REG_KEY]
         excluded_domains = registry.exclude_domains
 
         tracking: list[str] = []
         trackable: list[str] = []
+        single_state_type_key: set[tuple[str, str]] = set()
         for ent_id in entity_ids:
             ent_id_lower = ent_id.lower()
             domain = split_entity_id(ent_id_lower)[0]
             tracking.append(ent_id_lower)
             if domain not in excluded_domains:
                 trackable.append(ent_id_lower)
+            if domain in registry.state_group_mapping:
+                single_state_type_key.add(registry.state_group_mapping[domain])
+            elif domain == DOMAIN:
+                for group in iter(registry.group_entities):
+                    if group.entity_id == ent_id and group.single_state_type_key:
+                        single_state_type_key.add(group.single_state_type_key)
+                        break
+            else:
+                single_state_type_key.add((STATE_ON, STATE_OFF))
+
+        self.single_state_type_key = (
+            next(iter(single_state_type_key))
+            if len(single_state_type_key) == 1
+            else None
+        )
+        registry.group_entities.add(self)
```

Comment:
Can we avoid storing the entities in a shared container?  We try not to do that anymore

Reply:
Currently the source of truth for everything group is the state machine
https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/group.py

I'm not sure if its better to add another attribute or to register the `entity_id` in a `dict` of `entity_id -> `single_state_type_key` | None`

Reply:
The alternative is to enumerate them all the `ATTR_ENTITY_ID` values in the state machine for that entity id to figure out if its a single state type.  Its still faster than enumerating all the groups even if its not so great.  We also support multi level groups so I think we have to do it anyways

Reply:
Consider how `expand` expands groups https://github.com/home-assistant/core/blob/cdfd0aa7d4ac9bf46d151392e612e90d5d6f33da/homeassistant/helpers/template.py#L1248

Reply:
Now storing the state type for a group entity the same way as we do for a domain.